Refresh expired access tokens

studer-l opened the pull request:
Adds an additional flow to the existing middleware to refresh expired
tokens. The implementation handles multiple active grants. Expired
grants that failed to refresh are removed.

The refresh workflow is unconditionally activated. Since a token refresh
may occur for any request, access tokens are now always added to the
response's `:session`. This may break code which previously relied on the
`:session` only being set during the initial grant workflow. I do not
think this can be avoided.

If a refresh occurs, the tokens in `(:session request)` are left as-is,
the updated access tokens are accessibly via the existing
`:oauth2/access-token` key. This allows downstream handlers to observe
that a token refresh occurred.

There is a potential bug, where concurrent requests with expired tokens
may race. For example, consider a page containing a `css` and a `js`
resource. If a user's access token was to expire exactly as the
`index.html` finishes loading, their browser may concurrently fetch both
the `css` and `js` resources, triggering two concurrent token refresh
attempts with the same token, one of which may fail. I do not see a way to
address this without introducing considerable complexity.

I have added a timeout of 60 seconds to the refresh http request, which
means a slow oauth backend will cause users to become logged out. I
think this is more informative for users than hanging forever.

Fixes #40

Commit:
Refresh expired access tokens

Adds an additional flow to the existing middleware to refresh expired
tokens. The implementation handles multiple active grants. Expired
grants that failed to refresh are removed.

The refresh workflow is unconditionally activated. Since a token refresh
may occur for any request, access tokens are now always added to the
response's `:session`. This may break code which previously relied on the
`:session` only being set during the initial grant workflow. I do not
think this can be avoided.

If a refresh occurs, the tokens in `(:session request)` are left as-is,
the updated access tokens are accessibly via the existing
`:oauth2/access-token` key. This allows downstream handlers to observe
that a token refresh occurred.

There is a potential bug, where concurrent requests with expired tokens
may race. For example, consider a page containing a `css` and a `js`
resource. If a user's access token was to expire exactly as the
`index.html` finishes loading, their browser may concurrently fetch both
the `css` and `js` resources, triggering two concurrent token refresh
attempts with the same token, one of which may fail. I do not see a way to
address this without introducing considerable complexity.

I have added a timeout of 60 seconds to the refresh http request, which
means a slow oauth backend will cause users to become logged out. I
think this is more informative for users than hanging forever.

Fixes #40

weavejester commented:
Thanks for the PR. This might take a while for me to get through. Please ensure that the commit contains only changes relevant to this feature, that all lines are within 80 characters, and that only public functions have docstrings.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -114,26 +114,29 @@
-                 raise)))
+   (http/request
+    (-> (access-token-http-options profile
+                                   (request-params profile request))
```
Why has `request-params` been moved out here? I don't understand how this change connects to the token refreshes. Can you explain?

studer-l replied:
Ah I should have left a comment for this change, it's a bit confusing: I am trying to re-use `access-token-http-options` for the refresh http request as they share the same URL and credentials logic, except for the `form-params`, hence i extracted it as a parameter (in place of constructing those from the `request`).

If you feel this is convoluted, a possible alternative would be to `assoc` the `:form-params` in the caller, same as the `:async` is currently set from outside.

weavejester replied:
In which case, let's name things correctly. We can divide this into three functions:

1. `token-http-options` to contain the common code
2. `access-token-http-options` to operate as before
3. `refresh-token-http-options` to contain the options for refreshing tokens.

That way things are named correctly, and it makes the code a little more understandable later on.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
 
-(defn- assoc-access-tokens [request]
-  (if-let [tokens (-> request :session ::access-tokens)]
+(defn- get-expired
```
Generally speaking, we should avoid using 'get-' as part of pure function names. Instead `expired-access-tokens` would likely be a better name.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
-(defn- assoc-access-tokens [request]
-  (if-let [tokens (-> request :session ::access-tokens)]
+(defn- get-expired
+  "Returns expired profile keys and refresh tokens in `access-tokens`."
```
Docstrings on private functions should be avoided.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
+(defn- get-expired
+  "Returns expired profile keys and refresh tokens in `access-tokens`."
+  [access-tokens]
+  (let [now (new Date)]
```
Prefer `(Date.)` over `(new Date)`.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
+  (let [now (new Date)]
+    (for [[profile-key {:keys [expires refresh-token]}] access-tokens
+          :when (and expires refresh-token (.before expires now))]
+      {:profile-key profile-key :refresh-token refresh-token})))
```
Perhaps something like:

```clojure
(defn- expired-access-token? [[k {:keys [expires]}]]
  (and expires (.before expires (Date.))))

(defn- expired-access-tokens [access-tokens]
  (into {} (filter expired-access-token?) access-tokens)
```

That way we're just filtering the map and the output type is the same as the input type.

studer-l replied:
very good point! indeed this is just a filter

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
+       (http/request (comp respond format-access-token) raise))))
+
+(defn- valid-token? [token]
+  (and token (string? token) (not (str/blank? token))))
```
Do we need this check? I'm unsure of its purpose.

studer-l replied:
from my reading of the RFC, we should check whether the token is set, since it is optional for the authorizing server to provide one --- the other two predicates are... just garnish, and unlikely to do anything, so I did remove them and inline it

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
+(defn- valid-token? [token]
+  (and token (string? token) (not (str/blank? token))))
+
+(defn- refresh-all-tokens
+  "Refreshes all expired tokens, yielding an updated map of tokens"
+  ([profiles access-tokens]
+   (let [refresh-results
+         (for [{:keys [profile-key refresh-token]} (get-expired access-tokens)
+               :let [profile (profile-key profiles)]
+               :when (and profile (valid-token? refresh-token))]
+           [profile-key
+            (try (refresh-one-token profile refresh-token)
+                 (catch clojure.lang.ExceptionInfo _
+                   nil))])]
+     (reduce update-tokens access-tokens refresh-results)))
+  ([profiles access-tokens respond]
+   ;; strategy: launch all requests concurrently, keeping track of completed
+   ;; requests in `results`. When all requests have finished, respond.
+   (let [expired (get-expired access-tokens)
+         total (count expired)
+         results (atom {})  ;; map from profile-key to result
+         respond-when-done! #(when (= (count @results) total)
+                               (respond (reduce update-tokens access-tokens @results)))]
+     (if (zero? total)
+       (respond access-tokens)
+       (doseq [{:keys [profile-key refresh-token]} expired
+               :let [profile (profile-key profiles)]
+               :when (and profile (valid-token? refresh-token))]
+         (refresh-one-token profile refresh-token
+                            (fn [refresh-result]
+                              (swap! results assoc profile-key refresh-result)
+                              (respond-when-done!))
+                            (fn [_]
+                              (swap! results assoc profile-key nil)
+                              (respond-when-done!))))))))
```
It feels as if this function is attempting to do a lot, as we can see from it's length! I think we could split this up into three logical steps:

1. Collect the expired access tokens
2. Send them off via HTTP to be refreshed
3. Update the access token map

Only step 2 is different for sync/async. What we essentially want here is a way of running a bunch of async functions in parallel, and then to collect the results. That way we can have a function like:

```clojure
(defn- refresh-expired-tokens
  ([profiles access-tokens]
    (pmap (partial refresh-token profiles)
          (expired-access-tokens access-tokens)))
  ([profiles access-tokens respond raise]
    (amap (partial refresh-token profiles)
          (expired-access-tokens access-tokens)
          respond raise)))
```

If `map` (and `pmap`) expect a function `(f x)`, then `amap` would expect a function `(f x respond raise)`.

I'm running out of review time so I don't have enough time to sketch out the implementation of `amap`, but if I get time later I'll go ahead and do so.

studer-l replied:
Yes, you are right that this is a giant function; I usually refrain from writing novel combinators, but you are right that here I have an implementation of a `ring-style-async-map` that could be extracted.

I will give it a try later to see if I can simplify it.

I am not sure about the use of `pmap`, i do not think it is worth it to spawn a thread when in almost all cases only a single authentication provider is used.

studer-l replied:
I gave it a try, but ultimately decided not to implement a generic async map; there are a handful complications which I feel are not worth the headache of generalization (error handling, preserving order); the most recent commit adds a less ambitious, specialized version that works for associative containers and assigns `nil` to a key on failure. This does greatly simplify `refresh-all-tokens`  though, so I hope this addresses your concern :smile:

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
 
-(defn wrap-oauth2 [handler profiles]
+(defn wrap-oauth2
+  "Middleware that handles OAuth2 authentication flows.
+
+   Parameters:
+   * `handler`: The downstream ring handler
+   * `profiles`: A map of profiles
+
+   Each request URI is matched against the profiles to determine the appropriate
+   OAuth2 flow handler. If no match is found, the request is passed to the
+   downstream handler with existing access tokens added to the request under the
+  `:oauth2/access-tokens` key.
+
+  Expired tokens are refreshed using their refresh-token if possible. If refresh
+  fails, the access token is removed."
```
This docstring addition should be part of another commit ideally.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
-  (let [profiles  (for [[k v] profiles] (assoc v :id k))
-        launches  (into {} (map (juxt :launch-uri identity)) profiles)
-        redirects (into {} (map (juxt parse-redirect-url identity)) profiles)]
+  (let [id-profiles  (for [[k v] profiles] (assoc v :id k))
+        launches  (into {} (map (juxt :launch-uri identity)) id-profiles)
+        redirects (into {} (map (juxt parse-redirect-url identity)) id-profiles)]
```
This change appears to be cosmetic and can be omitted.

studer-l replied:
`id-profiles` is a sequence of profiles with an extra `id` key, `profiles` is a map used by the newly added code to facilitate looking up the expired profile based on the user's sessions.

weavejester replied:
Thanks, that makes sense.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,142 @@
            ((:redirect-handler profile (make-redirect-handler profile)) request)
-           (handler (assoc-access-tokens request)))))
+           (let [access-tokens (get-in request [:session ::access-tokens])
+                 refreshed-tokens (refresh-all-tokens profiles access-tokens)]
+             (-> request
+                 (assoc-access-tokens-in-request refreshed-tokens)
+                 handler
+                 (assoc-access-tokens-in-response refreshed-tokens))))))
```
What if there are no refreshed tokens?

studer-l replied:
i assume the naming was too confusing here --- the `refreshed-tokens` contain both the newly refreshed tokens as well as the existing, non-expired tokens.

weavejester reviewed (changes requested):
If I have time I'll do a more thorough review a little later, but these are my initial thoughts.

Commit:
reflow to 80 characters per line

Commit:
rename get-expired → expired-access-tokens

Commit:
remove docstrings on non-public functions; remove unrelated changes

Commit:
prefer `(Date.)` over `(new Date)`

Commit:
remove token type check

Commit:
refactor `expired-access-tokens` as filter

Commit:
fix confusing naming

Commit:
rename socket-timeout constant to be more specific

studer-l reviewed (commented):
Hi @weavejester thanks for the very prompt and thorough review :smile: 

I will see about doing some more work to refactor the giant function you mentioned, don't hesitate to let me know if anything else looks clunky :sweat_smile:

Commit:
refactor `refresh-all-tokens`

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,123 @@
-(defn- assoc-access-tokens [request]
-  (if-let [tokens (-> request :session ::access-tokens)]
+(defn- expired-access-tokens
+  [access-tokens]
```
We can omit the newline before the argument list if it all fits on one line, e.g.

```clojure
(defn- expired-access-tokens [access-tokens]
```

Is just as clear and uses less vertical whitespace.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,123 @@
-  (if-let [tokens (-> request :session ::access-tokens)]
+(defn- expired-access-tokens
+  [access-tokens]
+  (let [now (Date.)
+        expired-access-token? (fn [[_ {:keys [expires refresh-token]}]]
+                                (and refresh-token expires
+                                     (.before expires now)))]
```
I don't think we need to bother factoring the `now` out. It's not going to make much of a difference that I can see, and makes the code a little harder to parse. In addition, other parts of the code already use their own "now".

studer-l replied:
- extracting it leaves `expired-access-token?` a pure function
- there is a potentially confusing outcome where two tokens with identical expiration time take go different branches when expiration coincides with the evaluating of the `filter`

you are right that this has 0 practical impact, I will make the change tomorrow :smile:

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,123 @@
+        {:grant_type "refresh_token" :refresh_token refresh-token})
+       (assoc :socket-timeout refresh-socket-timeout)
+       http/request
+       format-access-token))
```
I think this would be clearer:

```clojure
(-> (http/request (refresh-token-http-options profile refresh-token))
    (format-access-token))
```

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,123 @@
          (if-let [profile (redirects uri)]
            ((:redirect-handler profile (make-redirect-handler profile)) request)
-           (handler (assoc-access-tokens request)))))
+           (let [access-tokens (->> (get-in request [:session ::access-tokens])
+                                    (refresh-all-tokens profiles))]
+             (-> request
+                 (assoc-access-tokens-in-request access-tokens)
+                 handler
+                 (assoc-access-tokens-in-response access-tokens))))))
```
This could probably be factored out into it's own function. So something like:

```clojure
(defn- wrap-refresh-access-tokens [handler profiles]
  (fn
    ([request] ...)
    ([request respond raise] ...)))
```

It should also only update the session if access tokens have expired or refreshed, otherwise the session should not be touched.

studer-l replied:
I am not sure updating the session in the response if changed is a good idea, as for users this might be rather surprising: If they previously relied on the `:session` key remaining unset, the breaking behavior only occurs when a token expires, which might be rather hard to debug.

But now that I think about this, we have to be careful as to not prevent users from implementing logout in `handler`. This could be addressed by checking if `:session` is explicitly set to `nil`, in which case we don't touch it.

weavejester replied:
The rules that govern the session middleware are:

- If the response `:session` key is missing the session is unaltered.
- If the response `:session` key is `nil`, the session is removed.
- If the response `:session` key is a map, it replaces the current session.

While it's possible to always replace the session with an identical result, it's not necessary and adds a session write command that may not be necessary. This might be a database update, for example. For that reason we need to only write the session if we want to alter it.

studer-l replied:
Thanks for the clarification; The performance overhead is a strong point and I have changed the behavior to only update the `:session` in the response if the access tokens have changed and if the handler did not already set the session.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,123 @@
+  [response tokens]
+  (if tokens
+    (assoc-in response [:session ::access-tokens] tokens)
+    response))
```
This assumes a `:session` key in the response map from the handler, which I don't believe is something you can rely on.

studer-l replied:
Ah, are you suggesting taking the `:session` from the `request` and merging the updated access tokens into that, if any?

studer-l replied:
(I simply went ahead and implemented this behavior now)

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,33 +191,123 @@
 
-(defn wrap-oauth2 [handler profiles]
+(defn wrap-oauth2
+  [handler profiles]
```
Newline can be omitted.

weavejester reviewed (changes requested):
Thank you very much for the updates! I have some additional feedback, but we're getting there. One thing that does need to be taken account of is what happens if the handler doesn't return a session. In that case, we should take the session from the request instead.

Commit:
refactor http param options

Commit:
remove spurious newlines in `defn`

Commit:
inline `now`

Commit:
slightly refactor `refresh-one-token` for clarity

Commit:
rework token refresh into a middle function

- extract `wrap-refresh-acceess-tokens`
- only set response `:session` if changed and if handler did not set it already
- add a test for the above

Commit:
minor refactor

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,102 @@
 
+(defn- assoc-access-tokens-in-response
+  [original-tokens updated-tokens response]
+  (if (and (not (contains? response :session))
```
This doesn't look quite correct. I believe it should be:

```clojure
(and (not (and (contains? response :session)
               (nil? (:session response))))
     (not= original-tokens updated-tokens))
```

Or more clearly:

```clojure
(and (not (nil-session? response))
     (not= original-tokens updated-tokens))
```

In that the only time we don't want to update the response is if the tokens haven't changed, or the session has been set explicitly to `nil` and is therefore being deleted.

studer-l replied:
Should be addressed in [26f2b34](https://github.com/weavejester/ring-oauth2/pull/56/commits/26f2b34d1176b5ef33b6b2814f4cbaa90bf1e5bf), with de morgan's law and branch order flipped

Commit:
update response session unless nil or tokens unchanged

ieugen commented:
hi @studer-l ,

I'm working on an app that uses oauth2 and I would need token refresh - since I have 5 min tokens. 
Do you have time to finish this PR? 
I'll see if I can implement and test it. 
Please ping me when you have anything new pushed.

Thanks,
Eugen

studer-l commented:
> hi @studer-l ,
> 
> I'm working on an app that uses oauth2 and I would need token refresh - since I have 5 min tokens. Do you have time to finish this PR? I'll see if I can implement and test it. Please ping me when you have anything new pushed.
> 
> Thanks, Eugen

Hi @ieugen 
Thank you for pinging me about this, sorry for letting this pull request linger; I should be able to find the time to finish this, no worries.

ieugen commented:
Hi, 

Anything I can do to help?
I've taken a break from this but I think I will get back on it in 2-3 weeks. 

Eugen

Commit:
cosmetic change

studer-l commented:
Hey @weavejester sorry to bother you with this again, I've gone through your comments again and I think they should all be addressed.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,103 @@
+
+(def refresh-socket-timeout 60000)
+
+(defn- refresh-token-request-options [profile refresh-token]
```
This should be called `refresh-token-http-options` for consistency with the other functions, no?

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,103 @@
+  (let [expired-access-token? (fn [[_ {:keys [expires refresh-token]}]]
+                                (and refresh-token expires
+                                     (.before expires (Date.))))]
+    (->> access-tokens (filter expired-access-token?) (into {}))))
```
`expired-access-token?` doesn't appear to use a closure, so we can instead use a private function rather than an inline one.

```clojure
(defn- expired-access-token? [{:keys [expires refresh-token]}]
  (and refresh-token expires (.before expires (Date.))))

(defn expired-access-tokens [tokens]
  (into {} (filter (comp expired-access-token? val)) tokens))
```

Same number of lines but clearer.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,103 @@
+      (respond {})
+      (doseq [[k v] m
+              :let [respond #(respond-when-done (swap! results assoc k %))
+                    raise (fn [_] (respond nil))]]
```
Probably better to use a normal `let` in this case, e.g.

```clojure
(doseq [[k v] m]
  (let [respond #(respond-when-done (swap! results assoc k %))
        raise   (fn [_] (respond nil))]]
    (f v respond raise)))
```

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,103 @@
+        (reduce update-tokens access-tokens)))
+  ([profiles access-tokens respond]
+   (async-map-values
+    (fn [[profile refresh-token] respond raise]
+      (refresh-one-token profile refresh-token respond raise))
```
In the synchronous `map`, the error catching function is in the map iteration function, but in `async-map-values`, the error catching function is built in.

My suggestion would be to make these two maps consistent. We assume that the mapping function, `f`, will not raise an exception:

```clojure
(defn- async-map-values [f respond m]
  (let [total (count m)
        results (atom {})
        respond-when-done #(when (= (count %) total) (respond %))]
    (if (zero? total)
      (respond {})
      (doseq [[k v] m]
        (f v #(respond-when-done (swap! results assoc k %)))))
```

We then do the equivalent of "catching" the exception when calling it in `refresh-all-tokens`:

```clojure
   (async-map-values
    (fn [[profile refresh-token] respond]
      (refresh-one-token profile refresh-token respond (fn [_] (respond nil)))
    (fn [refreshed-tokens]
      (respond (reduce update-tokens access-tokens refreshed-tokens)))
    (refresh-tasks profiles access-tokens))))
```

studer-l replied:
Indeed your refactor is more elegant, but its behavior differs from what I initially wrote. Assume two tokens require refresh, one of which fails:

* When following your suggestion, `refresh-all-tokens`'s continuation is called with `nil`.  So no refresh occurs.
* In the version you reviewed, the `results` in `async-map-values` will accumulate both a `nil` and a new access token, and the user's session will receive the one updated access token. The failed token will be retried on subsequent requests.

Do you have a preferred behavior in this case?

weavejester replied:
Probably the latter, but we need to extract the error handling from `async-map-values` .

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,103 @@
+                                           access-tokens updated-access-tokens)
+                   respond (comp respond update-session)]
+               (handler request respond raise)))]
+       (refresh-all-tokens profiles access-tokens respond)))))
```
We can make this a little more concise by making the variables a little shorter. Instead of `access-tokens` we can use `tokens` as there's no other sort of token that this function deals with. We can use the shorthand `'` suffix to indicate an updated version of a variable.

```clojure
(defn- wrap-refresh-access-tokens [handler profiles]
  (fn ([request]
       (let [tokens   (get-in request [:session ::access-tokens])
             tokens'  (refresh-all-tokens profiles tokens)
             request  (assoc-access-tokens-in-request request tokens')
             response (handler request)]
         (assoc-access-tokens-in-response token tokens' response)))
    ([request respond raise]
     (let [tokens (get-in request [:session ::access-tokens])]
       (refresh-all-tokens
        profiles tokens
        (fn [tokens']
          (let [request (assoc-access-tokens-in-request request tokens')
                respond #(respond (assoc-access-tokens-in-response
                                   % tokens tokens')))]
            (handler request respond raise)))))))
```

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -201,20 +296,21 @@
-  (let [profiles  (for [[k v] profiles] (assoc v :id k))
-        launches  (into {} (map (juxt :launch-uri identity)) profiles)
-        redirects (into {} (map (juxt parse-redirect-url identity)) profiles)]
+  (let [id-profiles  (for [[k v] profiles] (assoc v :id k))
```
Remove extra space after `id-profiles`.

weavejester reviewed (changes requested):
Thanks for the updated changes. I think we're nearly there. I've made a few more suggestions in order to improve code clarity.

Commit:
rename refresh-token-request-options -> refresh-token-http-options

Commit:
refactor expired-access-tokens

Commit:
replace `:let` with `let`

Commit:
remove extra space

Commit:
more concise `wrap-refresh-access-tokens`

chickendreanso commented:
Hi, any updates on this PR? It would be great to have the refresh token feature.

Commit:
correctly handle extra state from application

Correctly handles case where `request` contains `:session` map with extra data.

Commit:
format unit tests to 80 char width

studer-l commented:
Hey, apologies, I kind of forgot about this PR entirely!

I've had another look at the code and @weavejester 's review and I think I finally understood how to fix the extra `:session` state problem :sweat_smile: Please let me know if I am on the right track with my reasoning!

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,110 @@
+(defn- refresh-token-http-options [profile refresh-token]
+  (-> (token-http-options profile {:grant_type "refresh_token"
+                                   :refresh_token refresh-token})
+      (assoc :socket-timeout refresh-socket-timeout)))
```
Why is this a fixed value? Should it be a configurable value attached to the profile with 60000 as the default?

studer-l replied:
Yeah this could be configured in the profile, for consistency probably we should consider adding a timeout to the initial request as well i suppose --- but let's try and cut down the scope of this change and remove this timeout behavior for now. I think this timeout is only relevant for the case where multiple profiles are being refreshed concurrently, which i think is a niche usecase.

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,110 @@
+        (map (fn [[profile-key [profile refresh-token]]]
+               [profile-key
+                (try (refresh-one-token profile refresh-token)
+                     (catch clojure.lang.ExceptionInfo _ nil))]))
```
Can you explain why this is an `ExceptionInfo` and not an `Exception`? Do we only expect `ExceptionInfo` to be thrown from `http/request`?

studer-l replied:
I originally chose `ExceptionInfo` after experimenting with `clj-http`, but reading the code i can see that it can also throw various `java.io` and `java.net` exceptions --- mea culpa

weavejester commented on src/ring/middleware/oauth2.clj:
```diff
@@ -188,11 +191,110 @@
+    ;; handler did not change session, update original
+    :else (let [original-session (get request :session)]
+            (assoc response :session
+                   (assoc original-session ::access-tokens updated-tokens)))))
```
This `cond` statement feels like it could be simplified:

```clojure
(if (or (nil-session? response)
        (= original-tokens updated-tokens))
  response
  (let [session (get-current-session request response)]
    (assoc response :session (assoc session ::access-tokens updated-tokens))))
```

Where `get-current-session` is defined:

```clojure
(defn get-current-session [request response]
  (if (contains? response :session)
    (:session response)
    (:session request)))
```

Commit:
catch generic exception on sync refresh

Commit:
refactor `cond` in `assoc-access-tokens-in-response`

Commit:
remove timeout on refresh request

Commit:
extract error handling form `async-map-values`

Commit:
add a \n to test
